Add search, pagination, actions to all cache pages

Flynsarmy opened the pull request:
This PR contains a few changes:

* I fixed the errors running `composer phpunit` by adding [10up/wp_mock](https://packagist.org/packages/10up/wp_mock) as a dependency
* Added `phpunit` to the `composer test` script
* Added a Paginator class for use in our caches pages and tests for it
* Added search, pagination and actions to all caches pages

Flynsarmy committed:
Fix tests, Add tests to composer workflow, Add paginator class with tests, Add search, pagination, actions to all cache pages

Flynsarmy commented:
Fixing CI issues now.

Flynsarmy committed:
Fix CI errors

Flynsarmy commented:
While working on this PR I noticed there are tables such as crawl_cache with no `id` column with their primary key instead being the (quite long) md5 hash string. This presents a problem on the cache list pages because each checkbox in the list needs to pass a primary key back to the server when, for example, running the 'delete' action. This is currently done with GET and I'm unsure how the browser will respond with sending 200 of those long strings via GET.

* Should we add ID columns to every table that don't already have them?
* Are the MD5 hashes even necessary? If they're unique strings then it stands to reason that the unencrypted URL is also a unique string making the MD5 primary key redundant. Is there a reason we use them over standard integer IDs?

leonstafford commented:
Awesome, will review. 

It's been a while and a few changes since, but I believe the primary keys being hashes of urls/files was a performance/simplicity thing, as adding/checking caches is a very frequent action. 

@john-shaffer may have a fresher memory on these

leonstafford commented:
So, for now, I'd suggest keeping primary keys as they are, but should be no problem adding an id column for convenience of bulk deletion.

leonstafford commented:
And just to be aware of anywhere those caches are being used (look for references in source) and ensure there aren't any expecting column order or would fail to insert without the id. 

Partly due to static methods, but some end to end automated tests are really needed to be confident with these kind of changes. There's been a few flavours over the years, but I'm planning to use https://lokl.dev after next changes there to run such tests (locally and in CI environment). Due to nature of WP, being able to spin up/tear down a lot of different WordPress configurations to test against via browser is only way I feel confident with a lot of behaviours. 

The Static HTML Output repo has a lot of HTML parsing tests which are great for covering integration of quite a few classes and we'll end up with something similar here, too, but the detection/crawling/processing really needs those end to end tests.

Flynsarmy commented:
Yea I noticed a lot of testing is needed which is why I fixed them up in this PR and added tests for my new class. Will see how to go about adding tests for other classes in future PRs. I'll leave the id column problem for now. Too many other things that need to be worked on.

leonstafford commented:
Cool and sorry I've been absent from code for so long. Will help out once back into it!

john-shaffer commented:
The MD5 hashes are necessary because a URL can be up to 2083 characters, but MySQL won't allow unique indexes anywhere near that long. sgerlach added them in https://github.com/WP2Static/wp2static/pull/575. A PK id should be fine in addition.

leonstafford commented:
So @Flynsarmy do you want to try this PR without the ID? We'd be limited to like 2kb of data in GET/POST, which might not make it useful for bulk delete without the short ids?

john-shaffer commented:
I separated & added the phpunit changes in https://github.com/WP2Static/wp2static/commit/741c1568ceb000adc019b7d80b2f26580d935a7f. Changes to lock files shouldn't be accepted in PRs because it's a security risk [[1]](https://snyk.io/blog/why-npm-lockfiles-can-be-a-security-blindspot-for-injecting-malicious-modules/). Not that there's anything wrong, it's just best practices.

leonstafford commented:
Ah, good call, @john-shaffer. Do you still prefer to have them tracked in git? I kind of like that for reproducibility.

john-shaffer commented:
Yes, reproducibility is great.

Flynsarmy commented:
I'll add an id column to tables missing it and move the primary key over. I won't remove any columns and will make sure the old primaries are indexed appropriately with unique keys. Does this sound ok to you?

leonstafford commented:
Sounds good to me, thanks!

Flynsarmy committed:
Merge branch 'master' into paginatedPages

Flynsarmy commented:
I'm having some troubles adding the primary keys. If I change DeployCache for example to

```php
    public static function createTable() : void {
        global $wpdb;

        $table_name = $wpdb->prefix . 'wp2static_deploy_cache';

        $charset_collate = $wpdb->get_charset_collate();

        $sql = "CREATE TABLE $table_name (
            id mediumint(9) NOT NULL AUTO_INCREMENT,
            path_hash CHAR(32) NOT NULL,
            path VARCHAR(2083) NOT NULL,
            file_hash CHAR(32) NOT NULL,
            namespace VARCHAR(128) NOT NULL,
            PRIMARY KEY  (id),
            UNIQUE KEY path_hash_ns_idx (path_hash, namespace)
        ) $charset_collate;";

        require_once ABSPATH . 'wp-admin/includes/upgrade.php';
        dbDelta( $sql );
    }
```
which is correct syntax, `dbDelta` generates the following SQL queries:
```sql
ALTER TABLE wp_wp2static_deploy_cache CHANGE COLUMN `path_hash` path_hash CHAR(32) NOT NULL;
ALTER TABLE wp_wp2static_deploy_cache CHANGE COLUMN `path` path VARCHAR(2083) NOT NULL;
ALTER TABLE wp_wp2static_deploy_cache CHANGE COLUMN `file_hash` file_hash CHAR(32) NOT NULL;
ALTER TABLE wp_wp2static_deploy_cache CHANGE COLUMN `namespace` namespace VARCHAR(128) NOT NULL;
ALTER TABLE wp_wp2static_deploy_cache ADD COLUMN id mediumint(9) NOT NULL AUTO_INCREMENT;
ALTER TABLE wp_wp2static_deploy_cache ADD PRIMARY KEY  (`id`);
```
the problem is on the second last line. An SQL error is returned:
> ERROR 1075 (42000): Incorrect table definition; there can be only one auto column and it must be defined as a key

How do we go about solving this? Are we maintaining backwards compatibility so anyone can `git pull` or do we just expect breaking changes at the moment and recommend deleting all tables after this PR is merged?

Either way the plugin seems to need to be deactivated/reactivated for the `createTable` methods to fire.

leonstafford commented:
An aside, I haven't tried the new "Update in place" coming in WP 5.5 yet - wondering if it does deactivate/reactivate for us?

At this point, I think it's OK to make that breaking change. Let's take opportunity to also adjust version from 7.0-alpha to 7.0 or 7.1, which may alert those using it and likely to be pulling from source here. When next actual release is prepared, I'll do an email blast out and we can make a bit more noise about it. 

@john-shaffer's been looking to avoid/replace dBDelta in the future. 

Thinking out loud: As these aren't config tables holding anything critical, could we first check the table structure - if it's old, drop it and then recreate? Wouldn't want to be stuck using resources to check for it on each plugin usage, but could allow for non-interactive transition for the user, then we pull it out in future version?

john-shaffer commented:
It's okay to break things for now, but I already did the research so I'll put the results here. dbDelta is just horrible. Even if you learn all of its bugs and sadistic "quirks", the next SQL-fluent person that comes along is going to be in for some fun surprises when they try to change something. I don't really want to write a PHP schema migration library, but I think it's good to have a few functions&mdash;with logging&mdash;to handle the worst pain points with dbDelta.

We can make a variant of [Controller::ensure_index](https://github.com/WP2Static/wp2static/blob/741c1568ceb000adc019b7d80b2f26580d935a7f/src/Controller.php#L141-L174]) something like this:
```php
    public static function ensure_primary_key( string $table_name, string $field_name,
                                         string $create_pk_sql ) : bool {
        global $wpdb;

        $wpdb->query("ALTER TABLE $table_name DROP PRIMARY KEY");

        $query = $wpdb->prepare(
            "SHOW COLUMNS FROM $table_name WHERE Field = %s",
            $field_name
        );
        $columns = $wpdb->query( $query );

        if ( 0 === $columns ) {
            $result = $wpdb->query( $create_pk_sql );
            if ( false === $result ) {
                \WP2Static\WsLog::l( "Failed to create $field_name primary key on $table_name." );
            }
            return $result;
        } else {
            return true;
        }
    }
```

The `0 === $columns` condition might need to be changed.

and call with
```php
ensure_primary_key($table_name, $id, "ALTER TABLE $table_name ADD COLUMN id MEDIUMINT(9) NOT NULL AUTO_INCREMENT PRIMARY KEY");
```

As for requiring the plugin to be reactivated & deactivated, we can just add calls to `DeployCache::createTable` at the beginning of `ViewRenderer::renderDeployCache` and in https://github.com/WP2Static/wp2static/blob/6c3431ef64466aa83eed7b3a554eb090c2546052/src/Controller.php#L606-L611.

leonstafford commented:
@john-shaffer, I think you may have dropped this.... your cape! You're a hero!

Flynsarmy committed:
Add ID columns to cache tables. Use them for future backend page actions

Flynsarmy committed:
Fix CI errors

Flynsarmy commented:
I went with a solution that just removes the table if the ID column is missing then recreates so there is no manual deletion required - only a reactivation of the plugin. We can sort out some sort of process to not require plugin reactivations in another PR as this one is big enough already.

I originally tried to remove the primary key if the ID column isn't present then rely on dbDelta to set things up correctly but it just wouldn't work so this was the best I could manage.

After accepting this PR I'd recommend bumping the version number as mentioned in https://github.com/WP2Static/wp2static/pull/635#issuecomment-675295640

It helps we have the package marked as 'alpha' so issues like this are to be expected I guess.

leonstafford commented on src/Paginator.php:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    public function render() : void {
+        require __DIR__ . '/../views/_paginator.php';
```
Can we use the `WP2STATIC_PATH` from https://github.com/WP2Static/wp2static/blob/741c1568ceb000adc019b7d80b2f26580d935a7f/wp2static.php#L18 here?

leonstafford commented on src/Paginator.php:
```diff
@@ -0,0 +1,107 @@
+     * the array into keys and values, array_slice each then recombine
+     *
+     * @param array<mixed> $records
+     * @param int $page_size
```
FYI, I've not worried about including `@param`'s for parameters with strict types defined in the actual declaration, only for those that PHPStan complains about, like mixed or mixed arrays. No need to change anything here, but can save the extra lines in future if easier for you.

leonstafford commented on src/ViewRenderer.php:
```diff
@@ -111,8 +100,32 @@ public static function renderCrawlCache() : void {
 
-        $view = [];
-        $view['urls'] = CrawlCache::getURLs();
+        if ( ! empty( $_GET['action'] ) && ! empty( $_GET['id'] ) && is_array( $_GET['id'] ) ) {
```
I don't have it in this repo's PHPstan/phpcs rules yet, but we'll have checks soon to use `filter_input()` instead of directly accessing `_GET` and `_POST`. So using https://www.php.net/manual/en/function.filter-input.php should allow to remove all the `empty()` calls here.

leonstafford replied:
(not asking you to change all references like this if it's not being caught by PHPStan WP rules now. more of a heads up/suggestion)

leonstafford commented on src/ViewRenderer.php:
```diff
@@ -111,8 +100,32 @@ public static function renderCrawlCache() : void {
-        $view = [];
-        $view['urls'] = CrawlCache::getURLs();
+        if ( ! empty( $_GET['action'] ) && ! empty( $_GET['id'] ) && is_array( $_GET['id'] ) ) {
+            switch ( $_GET['action'] ) {
```
how about just an `if` here?

leonstafford reviewed (approved):
Brilliant, thanks do much for this very big effort, @Flynsarmy!

I added a few comments, but nothing blocking, so will GTM'd.
